crates/ruff_python_parser/src/parser/statement.rs

```diff
@@ -2346,84 +2428,238 @@ impl<'src> Parser<'src> {
+                    }
+
+                    parameters.kwarg = Some(Box::new(param));
+                    last_keyword_only_separator_range = None;
                 }
-                std::mem::swap(&mut args, &mut posonlyargs);
-            } else if parser.at(TokenKind::Name) {
-                let param = parser.parse_parameter_with_default(function_kind);
-                // Don't allow non-default parameters after default parameters e.g. `a=1, b`,
-                // can't place `b` after `a=1`. Non-default parameters are only allowed after
-                // default parameters if we have a `*` before them, e.g. `a=1, *, b`.
-                if param.default.is_none() && has_seen_default_param && !has_seen_asterisk {
-                    parser.add_error(
-                        ParseErrorType::DefaultArgumentError,
-                        parser.current_token_range(),
-                    );
+                TokenKind::Slash => {
+                    let slash_range = parser.current_token_range();
+                    parser.bump(TokenKind::Slash);
+
+                    if parameters.is_empty() {
+                        // test_err params_no_arg_before_slash
+                        // def foo(/): ...
+                        // def foo(/, a): ...
+                        parser.add_error(
+                            ParseErrorType::OtherError(
+                                "Position-only parameter separator not allowed as first parameter"
+                                    .to_string(),
+                            ),
+                            slash_range,
+                        );
+                    }
+
+                    if seen_positional_only_separator {
+                        // test_err params_multiple_slash_separator
+                        // def foo(a, /, /, b): ...
+                        // # def foo(a, /, b, c, /): ...
+                        parser.add_error(
+                            ParseErrorType::OtherError(
+                                "Only one '/' separator allowed".to_string(),
+                            ),
+                            slash_range,
+                        );
+                    }
+
+                    if seen_keyword_only_separator || parameters.vararg.is_some() {
+                        // test_err params_star_after_slash
+                        // def foo(*a, /): ...
+                        // def foo(a, *args, b, /): ...
+                        // def foo(a, *, /, b): ...
+                        // def foo(a, *, b, c, /, d): ...
+                        parser.add_error(
+                            ParseErrorType::OtherError(
+                                "'/' parameter must appear before '*' parameter".to_string(),
+                            ),
+                            slash_range,
+                        );
+                    }
+
+                    std::mem::swap(&mut parameters.args, &mut parameters.posonlyargs);
```

Comment:
What happens if there are two slashes? Does it then swap the parameters twice?

Reply:
Oh, this is a good point. This would then swap the parameters again which would create problem in the AST validation as the positional only parameters are visited first.

In this case, for better error recovery, I think we should extend the existing positional only arguments.

So, for the happy path where `/` is only once, we'd swap it but for any subsequent `/`, we'd extend it if there are any arguments.

Reply:
Hmm, no, we'd need to drop the token. For example,

```py
def foo(a, /, b, c, /): ...
```

Here, only `a` should be considered as positional-only argument while `b` and `c` are regular arguments otherwise the order would be incorrect. This should be the case at least with the current AST.